### PR TITLE
Add LIP: "Introduce Validators module"

### PR DESCRIPTION
**Pull request by ricott1**

### CODE REVIEW

File: proposals/lip-introduce_validators_module.md [hunk 1]
```diff
@@ -0,0 +1,577 @@
+```
+LIP: <LIP number>
+Title: Introduce validators module
+Author: Alessandro Ricottone <alessandro.ricottone@lightcurve.io>
+        Andreas Kendziorra <andreas.kendziorra@lightcurve.io>
+        Rishi Mittal <rishi.mittal@lightcurve.io>
+Type: Standards Track
```
Comment (IkerAlus) [inline proposals/lip-introduce_validators_module.md:7]: Add `Discussions-To:` field

File: proposals/lip-introduce_validators_module.md [hunk 2]
```diff
@@ -0,0 +1,577 @@
...
+
+
+### Ensuring Uniqueness of validator BLS Keys
+
+The validators module store maintains an account for each validator registered in the chain. In particular, it stores the [BLS key](https://github.com/LiskHQ/lips/blob/master/proposals/lip-0038.md) associated with the validator, used to sign commits. BLS keys have to be unique across the chain, i.e. two validators are not allowed to register the same BLS key. To easily check whether a BLS key has been previously registered, we use a _registered BLS keys substore_, to store all the registered validator BLS keys. Store keys are set to the BLS key and the corresponding store value to the address of the validator that registered the key. This allows to check for the existence of a certain BLS key in constant time.
+
+### Block Slots
+
+In the Lisk protocol, time is divided into intervals of fixed length. 
+These intervals are called a _block slots_ and at most one block can be added to the blockchain during each slot.
```
Comment (IkerAlus) [inline proposals/lip-introduce_validators_module.md:50]:
```suggestion
These intervals are called _block slots_ and at most one block can be added to the blockchain during each slot. 
```

File: proposals/lip-introduce_validators_module.md [hunk 3]
```diff
@@ -0,0 +1,577 @@
...
+### Ensuring Uniqueness of validator BLS Keys
+
+The validators module store maintains an account for each validator registered in the chain. In particular, it stores the [BLS key](https://github.com/LiskHQ/lips/blob/master/proposals/lip-0038.md) associated with the validator, used to sign commits. BLS keys have to be unique across the chain, i.e. two validators are not allowed to register the same BLS key. To easily check whether a BLS key has been previously registered, we use a _registered BLS keys substore_, to store all the registered validator BLS keys. Store keys are set to the BLS key and the corresponding store value to the address of the validator that registered the key. This allows to check for the existence of a certain BLS key in constant time.
+
+### Block Slots
+
+In the Lisk protocol, time is divided into intervals of fixed length. 
+These intervals are called a _block slots_ and at most one block can be added to the blockchain during each slot. 
+The length of a block slot is a constant called _block time_.
+The validator module provides the functions to convert timestamp to block slots and viceversa.
```
Comment (IkerAlus) [inline proposals/lip-introduce_validators_module.md:52]: suggestion:
`The validator module provides the functions to calculate the block slot given the timestamp and viceversa.`

File: proposals/lip-introduce_validators_module.md [hunk 4]
```diff
@@ -0,0 +1,577 @@
...
+Furthermore, it maintains the _generator list_, an array of addresses corresponding to the accounts that are eligible to generate a block in the current round. 
+Each entry of the generator list corresponds to a block slot.
```
Comment (IkerAlus) [inline proposals/lip-introduce_validators_module.md:54]: suggestion:
`Each entry of the generator list is assigned to a block slot.`

File: proposals/lip-introduce_validators_module.md [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+
+| Name          | Type    | Value       | Description |
+| ------------- |---------| ------------| ------------|
+|`MODULE_ID_VALIDATORS` | uint32 | TBD | ID of the validators module.|
+|`STORE_PREFIX_VALIDATORS_DATA` | bytes | 0x0000 | Store prefix of the validators data substore.|
+|`STORE_PREFIX_GENERATOR_LIST` | bytes | 0x4000 | Store prefix of the generator list substore.|
+|`STORE_PREFIX_BLS_KEYS` | bytes | 0x8000 | Store prefix of the registered BLS keys substore.|
+|`STORE_PREFIX_GENESIS_DATA` | bytes | 0xc000 | Store prefix of the genesis data substore.|
+|`INVALID_BLS_KEY` | bytes | 48 bytes all set to 0x00 | An invalid BLS key, used as a placeholder before a valid BLS key is registered.|
+|`BLOCK_TIME` | integer | Block time (in seconds) set in the chain configuration. |
```
Comment (IkerAlus) [inline proposals/lip-introduce_validators_module.md:74]: the description of this one is placed in the value column. My suggestion for value is to set `10 (value for the Lisk mainchain)`

**Review by IkerAlus — CHANGES_REQUESTED** [review]
Consider a couple of small fixes.

File: proposals/lip-introduce_validators_module.md (hunk 1 above)
Comment (ricott1) [reply proposals/lip-introduce_validators_module.md:7]: added

File: proposals/lip-introduce_validators_module.md (hunk 3 above)
Comment (ricott1) [reply proposals/lip-introduce_validators_module.md:52]: edited

File: proposals/lip-introduce_validators_module.md (hunk 4 above)
Comment (ricott1) [reply proposals/lip-introduce_validators_module.md:54]: edited

File: proposals/lip-introduce_validators_module.md (hunk 2 above)
Comment (ricott1) [reply proposals/lip-introduce_validators_module.md:50]: fixed

File: proposals/lip-introduce_validators_module.md (hunk 5 above)
Comment (ricott1) [reply proposals/lip-introduce_validators_module.md:74]: done

**Review by IkerAlus — APPROVED** [review]
Great work.

File: proposals/lip-introduce_validators_module.md [hunk 6]
```diff
@@ -0,0 +1,581 @@
...
+  properties: {
+    addresses: {
+      type: "array",
+      items: {
+        dataType: "bytes"
+      },
+      fieldNumber: 1
+    },
+    required: ["addresses"]
+}
```
Comment (Incede) [inline proposals/lip-introduce_validators_module.md:149]:
```suggestion
      },
    },
    required: ["addresses"]
}
```

**Review by Incede — CHANGES_REQUESTED** [review]

File: proposals/lip-introduce_validators_module.md [hunk 7]
```diff
@@ -480,8 +479,62 @@ This function returns a 20 bytes address.
...
+
+This function returns an object whose keys are generator addresses, with value corresponding to the number of block slots assigned to them in between the input timestamps.
+
+
+##### Execution
+
+```python
+getGeneratorBetweenTimestamps(startTimestamp, endTimestamp):
+    
+    if endTimestamp >= startTimestamp:
```
Comment (MaximeGagnebin) [inline proposals/lip-introduce_validators_module.md:511]: I think this inequality should be reversed. (and maybe we accept = ??)

File: proposals/lip-introduce_validators_module.md [hunk 8]
```diff
@@ -480,8 +479,62 @@ This function returns a 20 bytes address.
...
+        return invalid timestamps error
+
+    result = {}
+
+    startSlotNumber = floor((startTimestamp - genesisData.timestamp) / BLOCK_TIME)
+    endSlotNumber = floor((endTimestamp - genesisData.timestamp) / BLOCK_TIME)
+    totalSlots = endSlotNumber - startSlotNumber + 1
+
+    # Quick skip to assign directly many block slot to every generator in the list 
+    baseSlots = math.floor(totalSlots / generatorList.addresses.length)
```
Comment (MaximeGagnebin) [inline proposals/lip-introduce_validators_module.md:521]: no need for `math.` here.

File: proposals/lip-introduce_validators_module.md [hunk 9]
```diff
@@ -480,8 +479,62 @@ This function returns a 20 bytes address.
...
+    if baseSlots > 0:
+        totalSlots -= baseSlots * generatorList.addresses.length
+        for generatorAddress in generatorList:
```
Comment (MaximeGagnebin) [inline proposals/lip-introduce_validators_module.md:524]: -> in generatorList.addresses
?

File: proposals/lip-introduce_validators_module.md [hunk 10]
```diff
@@ -480,8 +479,62 @@ This function returns a 20 bytes address.
...
+            result[generatorAddress] = baseSlots
+
+
+    # Assign remaining slots
+    for slotNumber in range(startSlotNumber, startSlotNumber + totalSlots):
+        index = slotNumber % generatorList.addresses.length
```
Comment (MaximeGagnebin) [inline proposals/lip-introduce_validators_module.md:530]: slotIndex ? or use `index` in the following line.

**Review by MaximeGagnebin — CHANGES_REQUESTED** [review]

File: proposals/lip-introduce_validators_module.md (hunk 9 above)
Comment (ricott1) [reply proposals/lip-introduce_validators_module.md:528]: fixed

**Review by MaximeGagnebin — APPROVED** [review]
Looks all good to go !

**Review by Incede — APPROVED** [review]

**Review by shuse2 — APPROVED** [review]

File: proposals/lip-introduce_validators_module.md [hunk 11]
```diff
@@ -0,0 +1,648 @@
...
+        return False
+    if there exists an entry in the registered BLS keys substore with storeKey == blsKey:
+        return False 
+    if PopVerify(blsKey, proofOfPossession) != VALID:
+        return False
+
+    validatorAccount = {
+        generatorKey: generatorKey,
+        blsKey: blsKey,
+        proofOfPossession: proofOfPossession
```
Comment (Incede) [inline proposals/lip-introduce_validators_module.md:263]: This property doesn't exist in `validatorAccountSchema`, is it intended ?
```suggestion
```

**Review by Incede — CHANGES_REQUESTED** [review]

**Review by Incede — APPROVED** [review]

**Review by karmacoma — APPROVED** [review]

**Pull request merged**